### PR TITLE
Feature/model pruning

### PR DESCRIPTION
**Pull request by BradleyEdelman**
This PR introduces a **refined adaptive training strategy with a constant pruning ratio**. Key updates:

- **Score Calculation**: This version now computes an **accuracy score** and a **memory score** based on resource usage and model performance
- **Parameter Prioritization**: Accuracy and memory scores are weighted according to default or user-defined priority weighting to idenfity a priority list for parameter adjustment. Now, only the top priority paramater is adjusted in each epoch.
    - **Batch size priority** is weighted by memory usage.
    - **Learning rate priority** is inversely weighted by accuracy improvement (i.e. increases if accuracy stagnates).
- **Fixed Pruning Ratio**: Pruning is constant and is stripped at the end.
- **Code Quality Improvements**: Added pre-commit hooks and CI linting for consistency.

### CODE REVIEW

**Comment by BradleyEdelman** [comment]
Merge feature/model_pruning: Introducing a **refined adaptive training strategy with a constant pruning ratio** that utilizes a memory/accuracy scoring system and parameter priority weighting system to identify which parameters (batch size, learning rate) to adjust, how, and when.
